src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php

```diff
@@ -42,11 +46,16 @@ public function __construct(Connection $connection, SerializerInterface $seriali
     public function receive(callable $handler): void
     {
         while (!$this->shouldStop) {
-            $AMQPEnvelope = $this->connection->get();
-            if (null === $AMQPEnvelope) {
+            try {
+                $this->currentAmqpEnvelope = $this->connection->get();
```

Comment:
I'm not fond of the idea of forcing the receivers to have such local state at all. Can't you have another stamp (non-serialisable), AMQP-specific, that contains this `\AMQPMessage`? You'd call it `AmqpStamp` and that's it, you can just get it from the `Envelope`, no need of these `*currentMessage` methods anymore, just "normal" methods taking the `Envelope` as an argument.

Reply:
Indeed, I've gone back-and-forth on this. Your transport-specific stamp is a nice idea. It *would* leak this information out to userland, like middleware or event listeners. What do you think about that?

Also, in the latest commit, on retry, I re-send the Envelope for normal encoding/decoding. This means (and is by design) that if a new Envelope was created with new stamps, those stamps will be serialized & sent. It would mean that this AMQP-specific stamp *would* be serialized & sent. You mentioned "non-serialisable"... do you basically mean: give it a "sleep" method so that if/when we serialize it, it'll just be an empty object (i.e. it won't cause an error).

Reply:
Side note: if we did this and made methods like `reject()` require the `Envelope`, we wouldn't be able to handle the `MessageDecodingFailedException` in the `Worker` - catching that and rejecting would need to remain the responsibility of each transport, because there wouldn't be any `Envelope` that the `Worker` could send back to the `Receiver::reject()` method.

Reply:
> It would leak this information out to userland, like middleware or event listeners. What do you think about that?

That's a good thing IMHO. This allows users to go deep into customising it.

> This means (and is by design) that if a new Envelope was created with new stamps, those stamps will be serialized & sent.

Indeed, hence my (non-serializable) comment. We need a way to make them non-transportable, it's definitely a valid use-case. The `sleep` method works but is only about the `serialize` method... I'd imagine a `NonSerializableStampInterface` actually.

> if we did this and made methods like reject() require the Envelope, we wouldn't be able to handle the MessageDecodingFailedException in the Worker

Technically speaking we could actually create an empty envelope as part of the exception. It would make sense. But for the sake of this pull-request, let's keep it the responsibility of the transport :)

Reply:
I've made the change to use a stamp (`AmqpReceivedStamp`) and it's really, really nice - good suggestion. However, I also found out that serializing that stamp works just fine. The `AMQPEnvelope` inside it is just a simple object that doesn't cause any problems.

So, should we still add this `NonSerializableStampInterface`? Serializing the `AMQPEnvelope` doesn't cause any problems, as we're always looking for `$envelope->last(AMQPEnvelop::class)`. Also, to implement this, each serializer would actually need to create a *new* envelope with the `NonSerializableStampInterface` filtered out. Totally doable, but is this something we actually need to add?